### PR TITLE
Feature/wand

### PR DESCRIPTION
**Pull request by gnusi**
PR is the first step towards WAND support. It implements new iterator type for WAND support along with some optimizations and refactoring.

### CODE REVIEW

File: microbench/lower_bound_benchmark.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+///
+/// @author Andrey Abramov
+////////////////////////////////////////////////////////////////////////////////
+
+#include <benchmark/benchmark.h>
+#include <numeric>
+
+namespace {
+
+void BM_lower_bound(benchmark::State& state) {
```
Comment (MBkkt) [inline microbench/lower_bound_benchmark.cpp:28]: Btw interesting fact we can change asssimpotic to O(log(k)) where k is position in array
just change choose "m" strategy: 1 2 4...2^i and then find bigger value. You can use common binary search between last position and current

This algorithm more tolerant to value at the beginning and better in comparable of linear search for small founded index

File: core/formats/formats_10.cpp [hunk 2]
```diff
@@ -73,10 +73,14 @@ using namespace irs;
 // name of the module holding different formats
 constexpr string_ref MODULE_NAME = "10";
 
+template<bool Wand, uint32_t PosMin>
 struct format_traits {
   using align_type = uint32_t;
 
-  static constexpr uint32_t BLOCK_SIZE = 128;
+  static constexpr uint32_t block_size() noexcept { return 128; }
```
Comment (MBkkt) [inline core/formats/formats_10.cpp:80]: Why function?

File: core/formats/formats_10.cpp [hunk 3]
```diff
@@ -692,164 +766,176 @@ void postings_writer_base::end_term(version10::term_meta& meta, const uint32_t*
...
-    pay_->last = 0;
-  }
+    docs_.value.set(id);
 
-  ++docs_count_;
+    // first position offsets now is format dependent
+    pos_.last = FormatTraits::pos_min();
+    pay_.last = 0;
+  } else {
+    throw index_error(string_utils::to_string(
```
Comment (MBkkt) [inline core/formats/formats_10.cpp:882]: NIT you can use absl::StrCat or formatting

File: core/formats/formats_10.cpp [hunk 4]
```diff
@@ -1572,362 +1841,677 @@ class doc_iterator final : public irs::doc_iterator {
...
+  [[maybe_unused]] uint32_t notify{0};
+  while (this->begin_ < this->end_) {
+    doc.value += *this->begin_++;
 
-          last_level = level;
+    if constexpr (!IteratorTraits::position()) {
+      if (doc.value >= target) {
+        if constexpr (IteratorTraits::frequency()) {
+          this->freq_ = this->buf_.freqs + this->relative_pos();
+          assert((this->freq_ - 1) >= this->buf_.freqs && (this->freq_ - 1) < std::end(this->buf_.freqs));
```
Comment (MBkkt) [inline core/formats/formats_10.cpp:2481]: NIT separate asserts

File: core/formats/formats_10.cpp [hunk 5]
```diff
@@ -1572,362 +1841,677 @@ class doc_iterator final : public irs::doc_iterator {
...
+    assert(meta.freq);
 
-      if constexpr (IteratorTraits::frequency()) {
-        IteratorTraits::read_block(
-          *doc_in_,
-          enc_buf_,
-          doc_freqs_);
-      } else if constexpr (FieldTraits::frequency()) {
-        IteratorTraits::skip_block(*doc_in_);
+    if constexpr (IteratorTraits::position()) {
```
Comment (MBkkt) [inline core/formats/formats_10.cpp:2377]: NIT merge to single condition?

**Review by MBkkt — COMMENTED** [review]
End for today

File: core/formats/skip_list.hpp [hunk 6]
```diff
@@ -162,7 +162,7 @@ class SkipReaderBase : util::noncopyable {
   //////////////////////////////////////////////////////////////////////////////
   /// @returns number of elements in a skip-list
   //////////////////////////////////////////////////////////////////////////////
-  size_t NumLevels() const noexcept { return std::size(levels_); }
+  size_t NumLevels() const noexcept { return num_levels_; }
```
Comment (MBkkt) [inline core/formats/skip_list.hpp:165]: Is it optimization?

File: core/formats/formats_10.cpp [hunk 7]
```diff
@@ -2258,10 +2262,11 @@ class wanderator final : public doc_iterator_base<IteratorTraits, FieldTraits> {
       assert(skip_.NumLevels());
       assert(0 == prev_skip_.doc_ptr);
 
-      if (const doc_id_t skipped{skip_.Seek(target)};
-          skipped > (this->cur_pos_ + this->relative_pos())) {
+      const auto pos = this->cur_pos_;
+      this->cur_pos_ = skip_.Seek(target);
+
+      if (pos != this->cur_pos_) {
```
Comment (MBkkt) [inline core/formats/formats_10.cpp:2268]: looks like behavior change

File: core/formats/skip_list.hpp (hunk 6 above)
Comment (gnusi) [reply core/formats/skip_list.hpp:165]: yep ;-)

File: core/formats/formats_10.cpp (hunk 4 above)
Comment (gnusi) [reply core/formats/formats_10.cpp:2481]: Done

File: core/formats/skip_list.cpp [hunk 8]
```diff
@@ -35,288 +35,127 @@ namespace {
...
-    const memory_allocator& alloc /* = memory_allocator::global() */
-) {
-  max_levels = std::max(size_t(1), max_levels);
-  max_levels = std::min(max_levels, ::max_levels(skip_0_, skip_n_, count));
-  levels_.reserve(max_levels);
-  max_levels = std::max(max_levels, levels_.capacity());
+    const memory_allocator& alloc /* = memory_allocator::global() */) {
+  max_levels_ = std::min(
+    std::max(size_t{1}, max_levels),
+    ::max_levels(skip_0_, skip_n_, count));
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/formats/skip_list.cpp:55]: Nice catch! But I think it should be 
```
max_levels_ = std::clamp(max_levels, size_t{1}, ::max_levels(skip_0_, skip_n_, count));
```

File: core/formats/formats_burst_trie.cpp [hunk 9]
```diff
@@ -3398,6 +3351,19 @@ class field_reader final : public irs::field_reader {
...
 
+    virtual doc_iterator::ptr wanderator(
+        const seek_cookie& cookie,
+        IndexFeatures features) const override {
+#ifdef IRESEARCH_DEBUG
+      auto* impl = dynamic_cast<const ::cookie*>(&cookie);
+      assert(impl);
+#else
+      auto* impl = static_cast<const ::cookie*>(&cookie);
+#endif
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/formats/formats_burst_trie.cpp:3362]: not yet, but I'll implement

File: core/formats/columnstore2.cpp [hunk 10]
```diff
@@ -1008,7 +1008,9 @@ namespace columnstore2 {
 // -----------------------------------------------------------------------------
 
 void column::prepare(doc_id_t key) {
+#ifdef IRESEARCH_DEBUG
   assert(!sealed_);
+#endif
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/formats/columnstore2.cpp:1013]: Totally agree

File: core/formats/skip_list.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (MBkkt) [reply core/formats/skip_list.cpp:55]: Yes, you're right

File: core/formats/formats_burst_trie.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (gnusi) [reply core/formats/formats_burst_trie.cpp]: done

**Review by Dronplane — APPROVED** [review]
LGTM

File: core/formats/formats_10.cpp [hunk 11]
```diff
@@ -1570,364 +1930,725 @@ class doc_iterator final : public irs::doc_iterator {
...
+  assert(!IteratorTraits::position() || IteratorTraits::position() == FieldTraits::position());
+  assert(!IteratorTraits::offset() || IteratorTraits::offset() == FieldTraits::offset());
+  assert(!IteratorTraits::payload() || IteratorTraits::payload() == FieldTraits::payload());
+
+  // don't use doc_iterator for singleton docs
+  // must be ensured by the caller
+  assert(meta.docs_count > 1);
+  assert(this->begin_ == std::end(this->buf_.docs));
+
+  auto& term_state = static_cast<const version10::term_meta&>(meta);
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/formats/formats_10.cpp:2114]: `version10::term_meta` isn't polymorphic

File: core/formats/formats_10.cpp [hunk 12]
```diff
@@ -1570,364 +1930,725 @@ class doc_iterator final : public irs::doc_iterator {
...
+    this->refill();
+
+    // if this is the initial doc_id then
+    // set it to min() for proper delta value
+    doc.value += doc_id_t{!doc_limits::valid(doc.value)};
+  }
+
+  [[maybe_unused]] uint32_t notify{0};
+  while (this->begin_ != std::end(this->buf_.docs)) {
+    doc.value += *this->begin_++;
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/formats/formats_10.cpp:2195]: `begin_` is a pointer

File: core/formats/formats_10.cpp [hunk 13]
```diff
@@ -1537,27 +1698,226 @@ class position final : public irs::position,
...
+
+  void prepare(const term_meta& meta,
+               [[maybe_unused]] const index_input* pos_in,
+               [[maybe_unused]] const index_input* pay_in);
+
+  virtual attribute* get_mutable(irs::type_info::type_id type) noexcept override {
+    return irs::get_mutable(attrs_, type);
+  }
+
+  virtual doc_id_t seek(doc_id_t target) override {
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/formats/formats_10.cpp:1836]: we can cleanup this in other places in a separate PR

File: core/utils/iterator.hpp [hunk 14]
```diff
@@ -169,15 +170,11 @@ class ptr_iterator
   //////////////////////////////////////////////////////////////////////////////
   template<typename T>
   typename adjust_const<T>::reference as() const {
-    typedef typename std::enable_if<
-      std::is_base_of< base_element_type, T >::value, T
-    >::type type;
+    typedef typename std::enable_if_t<
+      std::is_base_of_v<base_element_type, T>, T
+    > type;
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/utils/iterator.hpp:175]: I think the whole `iterator.hpp` which is pretty outdated has to be adjusted at one :-)

File: core/formats/formats_10.cpp [hunk 15]
```diff
@@ -1537,27 +1698,226 @@ class position final : public irs::position,
...
+  uint32_t enc_buf_[IteratorTraits::block_size()]; // buffer for encoding
+  const doc_id_t* begin_{std::end(buf_.docs)};
+  uint32_t* freq_{}; // pointer into docs_ to the frequency attribute value for the current doc
+  index_input::ptr doc_in_;
+  doc_id_t left_{};
+};
+
+template<typename IteratorTraits, typename FieldTraits>
+void doc_iterator_base<IteratorTraits, FieldTraits>::refill() {
+  if (IRS_LIKELY(left_ >= IteratorTraits::block_size())) {
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/formats/formats_10.cpp:1760]: Hot path if `left_ >= IteratorTraits::block_size()`

File: core/formats/formats_10.cpp [hunk 16]
```diff
@@ -1570,364 +1930,725 @@ class doc_iterator final : public irs::doc_iterator {
...
+  if (!this->doc_in_) {
+    this->doc_in_ = doc_in->reopen(); // reopen thread-safe stream
+
+    if (!this->doc_in_) {
+      // implementation returned wrong pointer
+      IR_FRMT_ERROR("Failed to reopen document input in: %s", __FUNCTION__);
 
-    if (target <= doc.value) {
-      return doc.value;
+      throw io_error("failed to reopen document input");
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/formats/formats_10.cpp]: to ensure that log message is written

File: core/formats/formats_10.cpp [hunk 17]
```diff
@@ -1570,364 +1930,725 @@ class doc_iterator final : public irs::doc_iterator {
...
-      doc.value += *begin_++;
+    return;
+  }
+
+  auto skip_in = this->doc_in_->dup();
+
+  if (!skip_in) {
+    IR_FRMT_ERROR("Failed to duplicate input in: %s", __FUNCTION__);
+
+    throw io_error("Failed to duplicate document input");
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/formats/formats_10.cpp:2264]: to ensure that log message is written

File: core/formats/skip_list.cpp [hunk 18]
```diff
@@ -35,288 +35,126 @@ namespace {
...
-    uint64_t next_child = stream.file_pointer();
-    stream.write_vlong(child);
-    child = next_child;
-  }
 }
 
-void skip_writer::flush(index_output& out) {
-  const auto rend = levels_.rend();
+void SkipWriter::Flush(index_output& out) {
+  const auto rbegin = std::make_reverse_iterator(std::begin(levels_) + max_levels_);
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/formats/skip_list.cpp:70]: will address in the next PR

**Pull request merged**